collect_app/src/main/java/org/odk/collect/android/tasks/InstanceUploaderTask.java

```diff
@@ -60,68 +53,25 @@ protected void onPostExecute(Outcome outcome) {
-
-                        count -= selectionArgs.length - 1;
-                        selection.append(") and status=?");
-                        selectionArgs[i] = Instance.STATUS_SUBMITTED;
-
-                        try (Cursor results = new InstancesDao().getInstancesCursor(selection.toString(),
-                                selectionArgs)) {
-                            if (results != null && results.getCount() > 0) {
-                                List<Long> toDelete = new ArrayList<>();
-                                results.moveToPosition(-1);
-
-                                boolean isFormAutoDeleteOptionEnabled = settingsProvider.getGeneralSettings().getBoolean(GeneralKeys.KEY_DELETE_AFTER_SEND);
-
-                                // The custom configuration from the third party app overrides
-                                // the app preferences set for delete after submission
-                                if (deleteInstanceAfterSubmission != null) {
-                                    isFormAutoDeleteOptionEnabled = deleteInstanceAfterSubmission;
-                                }
-
-                                String formId;
-                                String formVersion;
-                                while (results.moveToNext()) {
-                                    formId = results.getString(results.getColumnIndex(InstanceColumns.JR_FORM_ID));
-                                    formVersion = results.getString(results.getColumnIndex(InstanceColumns.JR_VERSION));
-                                    if (InstanceUploaderUtils.shouldFormBeDeleted(formsRepository, formId, formVersion, isFormAutoDeleteOptionEnabled)) {
-                                        toDelete.add(results.getLong(results.getColumnIndex(InstanceColumns._ID)));
-                                    }
-                                }
-
-                                DeleteInstancesTask dit = new DeleteInstancesTask(instancesRepository, formsRepository);
-                                dit.execute(toDelete.toArray(new Long[toDelete.size()]));
-                            }
-                        } catch (SQLException e) {
-                            Timber.e(e);
-                        }
+                    Set<String> instanceIds = outcome.messagesByInstanceId.keySet();
+
+                    boolean isFormAutoDeleteOptionEnabled;
+
+                    // The custom configuration from the third party app overrides
+                    // the app preferences set for delete after submission
+                    if (deleteInstanceAfterSubmission != null) {
+                        isFormAutoDeleteOptionEnabled = deleteInstanceAfterSubmission;
+                    } else {
+                        isFormAutoDeleteOptionEnabled = settingsProvider.getGeneralSettings().getBoolean(GeneralKeys.KEY_DELETE_AFTER_SEND);
                     }
+
+                    Stream<Instance> instancesToDelete = instanceIds.stream()
```

Comment:
This is less efficient in terms of the amount of I/O performed as it makes a call for each instance that has just been submitted. I did think about optimizing this but realized:

* in most cases the instance count will be fairly low and so not a big deal
* if the instance count is large we'd actually prefer to do a query per instance rather than load them all into memory

Given that I think this is fine. If we really want we could add a way to grab several Instances by id but I suspect we could optimize this by holding onto instance refs we supposedly load during submission.